json/src/main/java/io/airlift/json/RecordAutoDetectModule.java

```diff
@@ -0,0 +1,63 @@
+package io.airlift.json;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.core.Version;
+import com.fasterxml.jackson.databind.AnnotationIntrospector;
+import com.fasterxml.jackson.databind.introspect.AnnotatedClass;
+import com.fasterxml.jackson.databind.introspect.VisibilityChecker;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+
+import java.lang.reflect.Modifier;
+
+public class RecordAutoDetectModule
+        extends SimpleModule
+{
+    @Override
+    public void setupModule(SetupContext context)
+    {
+        super.setupModule(context);
+        context.insertAnnotationIntrospector(new Introspector());
+    }
+
+    private static class Introspector
+            extends AnnotationIntrospector
+    {
+        private final VisibilityChecker.Std recordVisibilityChecker;
+
+        public Introspector()
+        {
+            recordVisibilityChecker = VisibilityChecker.Std.defaultInstance()
+                    .withGetterVisibility(JsonAutoDetect.Visibility.PUBLIC_ONLY)
+                    .withCreatorVisibility(JsonAutoDetect.Visibility.DEFAULT)
+                    .withGetterVisibility(JsonAutoDetect.Visibility.DEFAULT)
+                    .withFieldVisibility(JsonAutoDetect.Visibility.DEFAULT)
+                    .withIsGetterVisibility(JsonAutoDetect.Visibility.DEFAULT);
+        }
+
+        @Override
+        public VisibilityChecker<?> findAutoDetectVisibility(AnnotatedClass ac, VisibilityChecker<?> checker)
+        {
+            if (isRecord(ac.getRawType())) {
+                JsonAutoDetect overrideAnnotation = ac.getRawType().getAnnotation(JsonAutoDetect.class);
+                if (overrideAnnotation != null) {
+                    return VisibilityChecker.Std.construct(JsonAutoDetect.Value.from(overrideAnnotation));
+                }
+                return recordVisibilityChecker;
+            }
+            return checker;
+        }
+
+        @Override
+        public Version version()
+        {
+            return Version.unknownVersion();
+        }
+    }
+
+    // TODO replace with clazz.isRecord() once Airlift is at Java 17+
+    private static boolean isRecord(Class<?> clazz)
```

Comment:
Maybe we should do this by calling `Class.isRecord()` reflectively if it's available?

Reply:
@electrum I just pushed a change that uses a MethodHandle instead. wdyt?